server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -559,6 +499,94 @@ void phase1(IndexCommit snapshot, long startingSeqNo, IntSupplier translogOps, A
         }
     }
 
+    void recoverFilesFromSourceAndSnapshot(ShardRecoveryPlan shardRecoveryPlan,
```

Comment:
By my reading this wasn't running under `CancellableThreads#execute` before, but I think it deserves another call to `cancellableThreads.checkForCancel()` at the top anyway.